modules/indexer/code/zoekt/zoekt.go

```diff
@@ -0,0 +1,412 @@
+// Copyright 2025 The Gitea Authors. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+//go:build unix
+
+package zoekt
+
+import (
+	"bufio"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"slices"
+	"strconv"
+	"strings"
+
+	repo_model "code.gitea.io/gitea/models/repo"
+	"code.gitea.io/gitea/modules/analyze"
+	"code.gitea.io/gitea/modules/charset"
+	"code.gitea.io/gitea/modules/git"
+	"code.gitea.io/gitea/modules/indexer"
+	"code.gitea.io/gitea/modules/indexer/code/internal"
+	indexer_internal "code.gitea.io/gitea/modules/indexer/internal"
+	inner_zoekt "code.gitea.io/gitea/modules/indexer/internal/zoekt"
+	"code.gitea.io/gitea/modules/log"
+	"code.gitea.io/gitea/modules/setting"
+	"code.gitea.io/gitea/modules/typesniffer"
+
+	"github.com/go-enry/go-enry/v2"
+	"github.com/sourcegraph/zoekt"
+	"github.com/sourcegraph/zoekt/index"
+	"github.com/sourcegraph/zoekt/query"
+)
+
+type Indexer struct {
+	indexer_internal.Indexer // do not composite inner_zoekt.Indexer directly to avoid exposing too much
+	inner                    *inner_zoekt.Indexer
+	indexDir                 string
+}
+
+func (b *Indexer) SupportedSearchModes() []indexer.SearchMode {
+	return indexer.ZoektSearchModes()
+}
+
+func NewIndexer(indexDir string) *Indexer {
+	idxer := inner_zoekt.NewIndexer(indexDir)
+	return &Indexer{
+		Indexer:  idxer,
+		inner:    idxer,
+		indexDir: indexDir,
+	}
+}
+
+func newZoektIndexBuilder(indexDir string, repo *repo_model.Repository, targetSHA string) (*index.Builder, error) {
+	opts := index.Options{
+		IndexDir: indexDir,
+		SizeMax:  int(setting.Indexer.MaxIndexerFileSize),
+		IsDelta:  true,
+		RepositoryDescription: zoekt.Repository{
+			ID:   uint32(repo.ID),
+			Name: repo.FullName(),
```

Comment:
Not sure how the `zoekt.Repository` works, but what if a repo is renamed? Then the same repo ID with a different name?

Reply:
It appears that before and after renaming, search and indexing can both proceed normally: everything is associated through the repo ID. However, currently when deleting an index, it first looks up the repo name using the repo ID, and then deletes files on disk that have the repo name prefix. Therefore, there might be some residual data files from before the rename.


<img width="695" alt="image" src="https://github.com/user-attachments/assets/290a4cea-1620-471e-b5c9-928455483d14" />

adl/xx3 -- rename ->  adl/xx422

delete operation will only delete adl%2Fxx422_v16.00000.zoekt...

Perhaps I should check later if there is a way to directly delete an index using the repoid.